### ISSUE TITLE
mineral classification

### ISSUE THREAD

**Issue by ghost** [capstone-coal/pycoal#18]
we can use this issue to track progress on the mineral classification step. assigning to @heidiaclayton.

**Comment by heidiaclayton** [comment]
https://github.com/heidiaclayton/coal/tree/mineral-id-dev

**Comment by ghost** [comment]
Feel free to ping me this weekend. If I am available I will be willing to assist.

Although I am confident Spectral Python can do the job, as [mentioned](https://groups.google.com/forum/#!topic/coal-capstone/nXhRb1fA7lg) I would recommend checking out the [Orfeo Toolbox](https://www.orfeo-toolbox.org/) to see how it compares to Spy. [At first glance](https://www.orfeo-toolbox.org/features-2/) it has wider support for file formats as well as accelerated OpenCV machine learning algorithms than the Spy implementations. [Some](https://www.orfeo-toolbox.org/documentation/) of the Orfeo docs are in French, so let me know if I can help translate.

For comparison, here are the algorithms for dimensionality reduction:
http://www.spectralpython.net/algorithms.html#dimensionality-reduction
https://www.orfeo-toolbox.org/CookBook/Applications/app_DimensionalityReduction.html

And classification:
http://www.spectralpython.net/algorithms.html
https://www.orfeo-toolbox.org/CookBook/recipes/pbclassif.html

Like Spy, Orfeo has a NumPy [interface](https://www.orfeo-toolbox.org/CookBook/recipes/numpy.html).

In short, both libraries do some of the same things. I'm not sure how good the Python interface is for Orfeo. As long as the COAL API hides the details from the users, it would be straightforward to use one or the other.

It would be good to get feedback from @lewismc if you're leaning towards this option. We should also consider adding an evaluation of this to our Tech Review or whatever document it was that identified candidate libraries.

**Comment by heidiaclayton** [comment]
Here's the pull request: https://github.com/capstone-coal/coal/pull/23

**Comment by ghost** [comment]
When it comes time to optimize, an alternative to rewriting everything with Orfeo or another GPU-accellerated hyperspectral toolkit would be to reimplement Spy's perceptron classifier with OpenCV. This would be the sort of thing we could contribute back to Spectral Python.

**Comment by ghost** [comment]
@heidiaclayton and i will be pair programming on this beginning 2017-02-20, so i am adding myself as an assignee.

**Comment by heidiaclayton** [comment]
Hi @lewismc I just merged the code and am going to start with tests. Should I make a test directory and make a file called "mineral_test.py" or would you prefer another structure? Thanks.

**Comment by lewismc** [comment]
@heidiaclayton yes **tests** is absolutely fine. You'll see this defined [here](https://github.com/capstone-coal/pycoal/blob/master/setup.py#L40). mineral_test is grand as well. Thank you. Please create a separate issue and related pull request for the nosetests. Thanks.

**Comment by ghost** [comment]
@heidiaclayton to continue the conversation from issue https://github.com/capstone-coal/pycoal/pull/37, i wanted to get your thoughts on a "no data" classification value. presumably the output of this stage will be [an ENVI classified image](https://groups.google.com/forum/#!topic/coal-capstone/P40vdozJGHg), a matrix of integer classification values and a corresponding key.

my question is what value do we choose for pixels in which there is no data or in which the classification fails? our flightline is not perfectly rectangular, so values along the edges are represented by negative numbers. it would be convenient for mining identification if the output used **zero** as a no data value, and positive integers as class numbers.

something like this:
```
key = [ "no data", "class A", "class B", "class C", ... ]

image = [ [ 0, 0, 1, 1, 5, 3, 3, 0 ],
          [ 0, 1, 2, 5, 3, 7, 0, 0 ],
          [ 0, 5, 9, 5, 3, 6, 0, 0 ],
          [ 0, 1, 2, 5, 3, 6, 0, 0 ],
          [ 0, 1, 2, 7, 3, 7, 0, 0 ],
          [ 0, 1, 2, 5, 3, 7, 0, 0 ],
          [ 2, 2, 9, 5, 3, 0, 0, 0 ],
          [ 3, 1, 2, 5, 3, 0, 0, 0 ] ]
```

**Comment by heidiaclayton** [comment]
@browtayl I agree that 0 as a 'no data' value is good as long as the other classes are represented as positive integers.

**Comment by ghost** [comment]
thanks, that will be helpful for implementing mining id.

**Comment by ghost** [comment]
adding myself as an assignee again because i am working on this in pull https://github.com/capstone-coal/pycoal/pull/44

**Comment by ghost** [comment]
@heidiaclayton @lewismc initial mineral classification complete. tests to be implemented in https://github.com/capstone-coal/pycoal/issues/36. ASTER support to be implemented in issue https://github.com/capstone-coal/pycoal/issues/48. closed.

**Comment by lewismc** [comment]
boom

**Comment by ghost** [comment]
@lewismc i hope that doesn't mean i've broken something lol

**Comment by lewismc** [comment]
Nope, its still just the following syntax which is incorrect
```
Extracting pycoal-0.2-py2.7.egg to /home/travis/virtualenv/python2.7.12/lib/python2.7/site-packages

  File "/home/travis/virtualenv/python2.7.12/lib/python2.7/site-packages/pycoal-0.2-py2.7.egg/pycoal/mining.py", line 32

    return ???

           ^

SyntaxError: invalid syntax

  File "/home/travis/virtualenv/python2.7.12/lib/python2.7/site-packages/pycoal-0.2-py2.7.egg/pycoal/environment.py", line 26

    ???

    ^

SyntaxError: invalid syntax

  File "/home/travis/virtualenv/python2.7.12/lib/python2.7/site-packages/pycoal-0.2-py2.7.egg/pycoal/gis.py", line 75

    warpOptions = gdal.WarpOptions(format=???,

                                          ^

SyntaxError: invalid syntax
```

**Comment by ghost** [comment]
@lewismc understood. it's my priority to fix that as soon as possible in issues https://github.com/capstone-coal/pycoal/issues/19 and https://github.com/capstone-coal/pycoal/issues/20. you're also welcome to revert the merge that brought those in.

**Comment by lewismc** [comment]
nope, not at all. I'm working on the SDS so I have my hands full.

**Comment by ghost** [comment]
@lewismc very well. i am willing to revert the merge if you need `master` to build before i've finished mining and environment. it will be a few days before i have time to work on it.

**Comment by lewismc** [comment]
Nope all good. As you were. Good work.
